### ISSUE TITLE
Using GraphConvModel with UserDefinedFeaturizer and UserCSVLoader?

### ISSUE THREAD

**Issue by nole-lin** [deepchem/deepchem#1605]
**Description:**

Version '2.1.0'

I'm trying to add additional features to a GraphConvModel in addition to SMILES and I'm using UserDefinedFeaturizer along with UserCSVLoader to do so.

Here's what my CSV looks like:

0 | 1 | ... | 2303 | smiles | endpoint
-----------------------------------
0.2 | -0.07 | ... | 0.006 | B(B(N... | 1.0

And here is the code I'm using for that:

```
featurizer = dc.feat.UserDefinedFeaturizer(feature_fields=[str(x) for x in list(range(256*9))])
loader = dc.data.UserCSVLoader(tasks=["endpoint"], smiles_field="smiles", featurizer=featurizer)
dataset = loader.featurize('dataset.csv', shard_size=8192)

model = GraphConvModel(
        1, batch_size=10, mode='classification', dense_layer_size=128, tensorboard=True)
model.fit(dataset, nb_epoch=10)
```

However, I get this error when trying to fit the model:

```
AttributeError                            Traceback (most recent call last)
<ipython-input-42-fac9156287dc> in <module>
     17     model = GraphConvModel(
     18         1, batch_size=10, mode='classification', dense_layer_size=128, tensorboard=True)
---> 19     model.fit(dataset, nb_epoch=10)

~/anaconda3/envs/deepchemenv/lib/python3.6/site-packages/deepchem/models/tensorgraph/tensor_graph.py in fit(self, dataset, nb_epoch, max_checkpoints_to_keep, checkpoint_interval, deterministic, restore, submodel, **kwargs)
    151         self.default_generator(
    152             dataset, epochs=nb_epoch, deterministic=deterministic),
--> 153         max_checkpoints_to_keep, checkpoint_interval, restore, submodel)
    154 
    155   def fit_generator(self,

~/anaconda3/envs/deepchemenv/lib/python3.6/site-packages/deepchem/models/tensorgraph/tensor_graph.py in fit_generator(self, feed_dict_generator, max_checkpoints_to_keep, checkpoint_interval, restore, submodel)
    229                   self.session, n_enqueued, final_sample))
    230         enqueue_thread.start()
--> 231       for feed_dict in self._create_feed_dicts(feed_dict_generator, True):
    232         if self.queue_installed:
    233           # Don't let this thread get ahead of the enqueue thread, since if

~/anaconda3/envs/deepchemenv/lib/python3.6/site-packages/deepchem/models/tensorgraph/tensor_graph.py in _create_feed_dicts(self, generator, training)
   1069         yield {self._training_placeholder: train_value}
   1070     else:
-> 1071       for d in generator:
   1072         feed_dict = {}
   1073         for key, value in d.items():

~/anaconda3/envs/deepchemenv/lib/python3.6/site-packages/deepchem/models/tensorgraph/models/graph_models.py in default_generator(self, dataset, epochs, predict, deterministic, pad_batches)
    707           d[self.labels[0]] = y_b
    708         d[self.task_weights[0]] = w_b
--> 709         multiConvMol = ConvMol.agglomerate_mols(X_b)
    710         d[self.atom_features] = multiConvMol.get_atom_features()
    711         d[self.degree_slice] = multiConvMol.deg_slice

~/anaconda3/envs/deepchemenv/lib/python3.6/site-packages/deepchem/feat/mol_graphs.py in agglomerate_mols(mols, max_deg, min_deg)
    265     num_mols = len(mols)
    266 
--> 267     atoms_per_mol = [mol.get_num_atoms() for mol in mols]
    268 
    269     # Get atoms by degree

~/anaconda3/envs/deepchemenv/lib/python3.6/site-packages/deepchem/feat/mol_graphs.py in <listcomp>(.0)
    265     num_mols = len(mols)
    266 
--> 267     atoms_per_mol = [mol.get_num_atoms() for mol in mols]
    268 
    269     # Get atoms by degree

AttributeError: 'numpy.ndarray' object has no attribute 'get_num_atoms'
```

How do I fix this?

Thanks

**Comment by VIGS25** [comment]
`GraphConvModel` assumes you use a `ConvMol` object as your `X_b`. The `ConvMol` object contains several properties that lend to graph based machine learning applications.

You would need to change the `UserDefinedFeaturizer` to the `ConvMol` one

**Comment by nole-lin** [comment]
OK if I replace `UserDefinedFeaturizer` with `ConvMolFeaturizer` it works. But what if I want to include both graph based features as well as user-defined features?

**Comment by VIGS25** [comment]
There's an option to add user defined features for ConvMol featurizer.

**Comment by Jiang-Lu-CAS** [comment]
> There's an option to add user defined features for ConvMol featurizer. You can supply that under atom_properties argument.

Hi，how can I add the atom_properties？ I can't add it properly. 
`featurizer = ConvMolFeaturizer(atom_properties=[prop_name])`
what's the prop_name? How can I find the list and write it in a right way. Thank you.

**Comment by filipm90** [comment]
Hi,

I have the same question: I have SMILES which I want to featurize using ConvMolFeaturizer(), but also concatenate with custom features that I pre-computed and I know they can add value. How can I do this? And this is not solved by atom_properties within ConvMolFeaturizer.

Thanks,
Filip

**Comment by rbharath** [comment]
@filipm90 This is unfortunately not easy to do right now with DeepChem. As a question though, are your features per-atom or per-molecule? For per-atom it might be possible for us to get something together that works, but per-molecule is trickier (since there's no obvious way to feed into the GraphConv architecture

**Comment by filipm90** [comment]
@rbharath No, unfortunately these are per compound features. There is an implementation of combining convolutions with custom-made features in ChemProp, and I was keen to explore if this could be implemented in DeepChem. Alternatively, as I talked with one of my colleagues, one needs to do two separate NNs, one for convolutions and other regular for custom features and then concatenate at the end. It would be great to have a tutorial, or implementation for this, as I assume that many would like to combine convolutions with, e.g., in silico prediction of other endpoints. Thank you!

**Comment by rbharath** [comment]
@filipm90 Would you mind pointing me at the Chemprop implementation? I think your colleague's suggestion for doing two separate NNs is sensible. You could make a custom Keras model to do this and wrap it as a DeepChem model. This would definitely make for a nice tutorial. Let me mark this as a potential topic for contribution!

**Comment by filipm90** [comment]
I discussed with colleague as he's more familiar with ChemProp.

He pointed the features generators here:
https://github.com/chemprop/chemprop/blob/master/chemprop/features/features_generators.py
 
Which is then used to construct a Molecule object (lines 77-90):
https://github.com/chemprop/chemprop/blob/master/chemprop/data/data.py

But apparently it's harder to dig the actual network implementation of featurizer.

@rbharath Thanks for taking it in consideration and suggesting to have a tutorial for this (and possibly implementation). In some of our work we've seen an improvement in concatenating features and it would be worth adding this to DeepChem.

**Comment by rbharath** [comment]
Yes agreed this would be a good feature/tutorial. I don't have bandwidth to take this on myself at present, but as a thought, if we get Google summer of code again this year, this might be a good student project in case no one has implemented it by then. Will keep this issue open to track!

Alternatively, if anyone is interested to take on the implementation task, please drop a note here and go for it!

**Comment by djbalin** [comment]
Hi @rbharath, I'm reviving this old thread as it discusses exactly what I am trying to do in my master's thesis: concatenating my own feature vector onto the ones generated by DeepChem. Ideally, I want to append _both_ nodewise features as well as a graph/molecule-wise feature, but I am struggling to find the infrastructure that allows me to do this. 
I was able to simply append a column vector onto the `node_features` of e.g. `DataSet.X` to handle the nodewise features, and, for the molecule-wise features, I _think_ that I am able to use a GraphGather layer and then append one dimension onto the output provided by that. However, I have not been able to combine the two approaches and use them with a model.

What do you think - is this possible with DeepChem? Perhaps you implemented this since your discussion in this thread, but I haven't been able to find it :)

Thank you very much!
